supportedBackends/spark/src/main/scala/cromwell/backend/impl/spark/SparkJobExecutionActor.scala

```diff
@@ -0,0 +1,207 @@
+package cromwell.backend.impl.spark
+
+import java.nio.file.FileSystems
+import java.nio.file.attribute.PosixFilePermission
+
+import akka.actor.Props
+import cromwell.backend.BackendJobExecutionActor.{BackendJobExecutionResponse, FailedNonRetryableResponse, SucceededResponse}
+import cromwell.backend.impl.spark.SparkClusterProcess._
+import cromwell.backend.io.JobPaths
+import cromwell.backend.sfs.{SharedFileSystem, SharedFileSystemExpressionFunctions}
+import cromwell.backend.{BackendConfigurationDescriptor, BackendJobDescriptor, BackendJobExecutionActor}
+import cromwell.core.{TailedWriter, UntailedWriter}
+import wdl4s.parser.MemoryUnit
+import wdl4s.util.TryUtil
+
+import scala.concurrent.{Future, Promise}
+import scala.sys.process.ProcessLogger
+import scala.util.{Failure, Success, Try}
+import scala.language.postfixOps
+
+object SparkJobExecutionActor {
+  val fileSystems = List(FileSystems.getDefault)
+
+  def props(jobDescriptor: BackendJobDescriptor, configurationDescriptor: BackendConfigurationDescriptor): Props =
+    Props(new SparkJobExecutionActor(jobDescriptor, configurationDescriptor))
+}
+
+class SparkJobExecutionActor(override val jobDescriptor: BackendJobDescriptor,
+                             override val configurationDescriptor: BackendConfigurationDescriptor) extends BackendJobExecutionActor with SharedFileSystem {
+
+  import SparkJobExecutionActor._
+  import better.files._
+  import cromwell.core.PathFactory._
+  import SparkClusterConstants._
+  import SparkCommands._
+
+  private val tag = s"SparkJobExecutionActor-${jobDescriptor.key.tag}:"
+
+  lazy val cmds = new SparkCommands
+  lazy val clusterExtProcess = new SparkClusterProcess()(context.system)
+
+  lazy val extProcess = new SparkProcess{}
+  lazy val clusterManagerConfig = configurationDescriptor.backendConfig.getConfig("cluster-manager")
+  private val fileSystemsConfig = configurationDescriptor.backendConfig.getConfig("filesystems")
+  private val sparkMaster = configurationDescriptor.backendConfig.getString("master").toLowerCase
+  private val sparkDeployMode = configurationDescriptor.backendConfig.getString("deployMode").toLowerCase
+  override val sharedFileSystemConfig = fileSystemsConfig.getConfig("local")
+  private val workflowDescriptor = jobDescriptor.descriptor
+  private val jobPaths = new JobPaths(workflowDescriptor, configurationDescriptor.backendConfig, jobDescriptor.key)
+
+  // Files
+  private val executionDir = jobPaths.callRoot
+  private val scriptPath = jobPaths.script
+
+  private lazy val stdoutWriter = extProcess.untailedWriter(jobPaths.stdout)
+  private lazy val stderrWriter = extProcess.tailedWriter(100, jobPaths.stderr)
+
+  private lazy val clusterStdoutWriter = clusterExtProcess.untailedWriter(jobPaths.stdout)
+  private lazy val clusterStderrWriter = clusterExtProcess.tailedWriter(100, jobPaths.stderr)
```

Comment:
It feels like the deploy mode is an abstraction that needs to be broken out here.  There are parallel groups of writers and processes, and "if cluster mode" logic which would seem to benefit from encapsulation and polymorphism.